File: builtins/FinalizationRegistry.js
// Placeholder: FinalizationRegistry
//
// Full implementation requires a native Rust FinalizationRegistry backing
// (gc-observation hook + cleanup callback queue + job integration).
// Until then, this stub satisfies the global name so that test262 harness
// dependencies that reference `FinalizationRegistry` do not throw ReferenceError.
//
// See: ECMA-262 §27.2 — FinalizationRegistry Objects

var ops = __ops__;
var ThrowTypeError = ops.ThrowTypeError;

// Stub constructor — no-op until native backing lands.
var FinalizationRegistry = function FinalizationRegistry(cleanupCallback) {};
FinalizationRegistry.prototype.constructor = FinalizationRegistry;

FinalizationRegistry.prototype.cleanupSome = function FinalizationRegistryCleanupSome(callback) {
  if (this === null || this === undefined) throw ThrowTypeError("FinalizationRegistry.prototype.cleanupSome called on null or undefined");
  // No-op until native backing lands.
};

FinalizationRegistry.prototype.register = function FinalizationRegistryRegister(target, heldValue, unregisterToken = undefined) {
  if (this === null || this === undefined) throw ThrowTypeError("FinalizationRegistry.prototype.register called on null or undefined");
  // No-op until native backing lands.
};

FinalizationRegistry.prototype.unregister = function FinalizationRegistryUnregister(unregisterToken = undefined) {
  if (this === null || this === undefined) throw ThrowTypeError("FinalizationRegistry.prototype.unregister called on null or undefined");
  return false;
};

File: crates/quench-runtime/src/builtin_meta/finalization_registry.rs
//! FinalizationRegistry intrinsic metadata.

use crate::ops::Builtin;

pub fn property(key: &str) -> Option<Builtin> {
    use Builtin::*;
    match key {
        "constructor" => Some(FinalizationRegistry),
        "register" => Some(FinalizationRegistryRegister),
        "unregister" => Some(FinalizationRegistryUnregister),
        _ => None,
    }
}

pub const fn fn_name(builtin: Builtin) -> Option<&'static str> {
    use Builtin::*;
    Some(match builtin {
        FinalizationRegistryRegister => "FinalizationRegistry.prototype.register",
        FinalizationRegistryUnregister => "FinalizationRegistry.prototype.unregister",
        _ => return None,
    })
}

pub const fn fn_len(builtin: Builtin) -> Option<f64> {
    use Builtin::*;
    Some(match builtin {
        FinalizationRegistryRegister => 2.0,
        FinalizationRegistryUnregister => 1.0,
        _ => return None,
    })
}

pub const fn short_name(builtin: Builtin) -> Option<&'static str> {
    use Builtin::*;
    Some(match builtin {
        FinalizationRegistryRegister => "register",
        FinalizationRegistryUnregister => "unregister",
        _ => return None,
    })
}
